### PR TITLE
fix: `return $this` to `static` return type

### PR DESCRIPTION
**Pull request by Zales0123**
Hello 👋 

First of all, thank you to all the creators and contributors for this amazing tool 🖖 I've been using it extensively for some legacy code refactoring and it helps a lot.
One of the things I noticed during my last work was the usage of `@return $this` doc-block instead of `@return self` or `@return static` I usually used before PHP8. Apparently, it's [quite common indeed](https://github.com/search?q=%22%40return+%24this%22+language%3APHP+repo%3Asymfony%2Fsymfony&type=code) :dancer: The current version of `PhpdocToReturnTypeFixer` does not support this case, so I've tried to fix this problem.

As it's my first contribution to this project, I would be grateful for any comments or guidelines 🫡
Cheers 🚀

### CODE REVIEW

**Comment by Wirone** [comment]
Hi Mateusz, nice to see you 😁. Thanks for the nice contribution!

It would be great if this case was covered in [this integration test](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/tests/Fixtures/Integration/priority/phpdoc_to_return_type%2Cno_superfluous_phpdoc_tags.test), because if there's native return type, then `@return $this` is superfluous.

Also I am not sure if `$this` should be always converted to `static` - if class is final, then it rather should be `self`. I don't remember if we have analyzer for checking class modifiers, but surely there are fixers that do it so you can check how it's done 🙂.

**Comment by Zales0123** [comment]
Thanks for your valuable feedback, I will try to push it forward tomorrow 🫡

File: src/AbstractPhpdocToTypeDeclarationFixer.php [hunk 1]
```diff
@@ -185,6 +185,10 @@ protected function getCommonTypeFromAnnotation(Annotation $annotation, bool $isR
             return null;
```
Comment (keradus) [inline src/AbstractPhpdocToTypeDeclarationFixer.php:185]: Hey,
I would like to consider this change to be a feature, not bugfix.
ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/feature-or-bug.rst#feature

**Comment by keradus** [comment]
> It would be great if this case was covered in [this integration test](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/tests/Fixtures/Integration/priority/phpdoc_to_return_type%2Cno_superfluous_phpdoc_tags.test), because if there's native return type, then @return $this is superfluous.

it is not.
`@return $this` is giving _more_ information than `: static`, thus docblock is not superfluous

File: tests/Fixer/FunctionNotation/PhpdocToReturnTypeFixerTest.php [hunk 2]
```diff
@@ -394,6 +394,20 @@ function bar() {}
             '<?php /** @return int */ fn(): int => 1;',
             '<?php /** @return int */ fn() => 1;',
         ];
+
+        yield '$this accessor' => [
```
Comment (keradus) [inline tests/Fixer/FunctionNotation/PhpdocToReturnTypeFixerTest.php:398]: Hey @Zales0123 ,
I let myself to update your testcase - now, it executes same test on PHP 8+, but also verifies that there is no change when executing below PHP 8.

unfortunately, your change is causing valid PHP 7.4 code to crash after changes applied by Fixer on PHP 7.4 runtime

**Comment by Wirone** [comment]
> `@return $this` is giving more information than `: static`, thus docblock is not superfluous

OK, you're right @keradus. `@return $this` is a `static` narrowed to "exact same instance". Personally I don't use it because I don't see value in it, hence my initial comment 😉. Can you share real life example where it makes a difference?

**Comment by keradus** [comment]
```php
$foo = new();
$foo->setX(a)->setY(x)->setZ(a)
```

but in the end, the valid use case you would love to see doesn't make any difference for me, to be honest. if it's different, it's different - let's keep it different.

**Pull request closed without merging**